osutil/sync_dir.go

```diff
@@ -0,0 +1,158 @@
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
+// returned to the caller for any extra processing that might be required (e.g.
+// to run some helper program).
+//
+// The return value is: created, corrected, removed []string, err error
+func EnsureDirState(dir, glob string, content map[string]*FileState) ([]string, []string, []string, error) {
+	matches, err := filepath.Glob(path.Join(dir, glob))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	found := make(map[string]bool)
+	var created, corrected, removed []string
+	// Analyze files that inhabit the subset defined by our glob pattern.
+	for _, name := range matches {
+		baseName := path.Base(name)
+		// Remove files that should not be here.
+		var expected *FileState
+		var shouldBeHere bool
+		if expected, shouldBeHere = content[baseName]; !shouldBeHere {
+			if err := os.RemoveAll(name); err != nil {
+				return created, corrected, removed, err
+			}
+			removed = append(removed, baseName)
+			continue
+		}
+		var file *os.File
+		if file, err = os.OpenFile(name, os.O_RDWR, 0); err != nil {
+			return created, corrected, removed, err
+		}
+		defer file.Close()
+		var stat os.FileInfo
+		if stat, err = file.Stat(); err != nil {
+			return created, corrected, removed, err
+		}
+		found[baseName] = true
+		// Check that file has the right content
+		needsRewrite, err := fileNeedsRewrite(file, stat, expected)
+		if err != nil {
+			return created, corrected, removed, err
+		}
+		if needsRewrite {
+			if err := AtomicWriteFile(file.Name(), expected.Content, expected.Mode, 0); err != nil {
+				return created, corrected, removed, err
+			}
+			corrected = append(corrected, baseName)
+			// NOTE: rewriting files also fixes permissions so we can skip the last stage
+			continue
+		}
+		// Check that file has the right meta-data
+		changed := false
+		currentPerm := stat.Mode().Perm()
+		expectedPerm := expected.Mode.Perm()
+		if currentPerm != expectedPerm {
+			if err := file.Chmod(expectedPerm); err != nil {
+				return created, corrected, removed, err
+			}
+			changed = true
+		}
+		if st, ok := stat.Sys().(*syscall.Stat_t); ok {
+			if st.Uid != expected.UID || st.Gid != expected.GID {
+				if err := file.Chown(int(expected.UID), int(expected.GID)); err != nil {
+					return created, corrected, removed, err
+				}
+				changed = true
+			}
+		}
+		if changed {
+			corrected = append(corrected, baseName)
+		}
+	}
+	// Create files that were not found but are expected
+	for baseName, expected := range content {
+		if baseName != path.Base(baseName) {
+			err := fmt.Errorf("expected files cannot have path component: %q", baseName)
```

Comment:
This is a serious bug.. the call site broke the contract of the function, which means it cannot possibly work. I suggest panicking for the time being:
```
panic(fmt.Sprintf("EnsureDirState got a filename which is actually a file path: %q", baseName))
```